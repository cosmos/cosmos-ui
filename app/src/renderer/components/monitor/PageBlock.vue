<template lang="pug">
tm-page(:title="pageBlockTitle")
  tm-data-loading(v-if="blockchain.blockLoading")
  tm-data-empty(v-else-if="!block || !block.header || !blockMeta")
  template(v-else)
    div(slot="menu"): tool-bar
      router-link(:to="{ name: 'block', params: { block: block.header.height - 1 }}"
        :disabled="block.header.height === 0"
        :event="block.header.height === 0 ? '' : 'click'"
        v-tooltip.bottom="'Older Block'")
        i.material-icons chevron_left
      router-link(:to="{ name: 'block', params: { block: block.header.height + 1 }}"
        :disabled="!nextBlockAvailable"
        :event="nextBlockAvailable ? 'click' : ''"
        v-tooltip.bottom="'Newer Block'")
        i.material-icons chevron_right
    tm-block(
      :blockMeta="blockMeta",
      :blockHeaderTime="blockHeaderTime",
      :block="block",
      :loading="blockchain.blockLoading",
      :txs="txs",
      :currentUser="wallet.address")
</template>

<script>
import moment from "moment"
import { mapGetters } from "vuex"
import num from "scripts/num"
import ToolBar from "common/TmToolBar"
import { TmPage, TmDataEmpty, TmDataLoading, TmBlock } from "@tendermint/ui"
export default {
  name: "page-block",
  components: {
    TmDataLoading,
    TmDataEmpty,
    ToolBar,
    TmPage,
    TmBlock
  },
  computed: {
    ...mapGetters(["blockchain", "blockTxInfo", "config", "wallet"]),
    blockchainHeight() {
      return this.blockchain.blocks.length > 0
        ? this.blockchain.blocks[0].header.height
        : 0
    },
    txs() {
      return (
        this.blockTxInfo &&
        this.blockTxInfo.map(tx => {
          tx.formattedTime = moment(tx.time).format("MMMM Do YYYY — h:mm:ss a")
          return tx
        })
      )
    },
    block() {
      return this.blockchain.block
    },
    blockHeaderTime() {
      if (this.block.header) {
        return moment(this.block.header.time).format("MMMM Do YYYY — hh:mm:ss")
      } else {
        return "Loading..."
      }
    },
    blockMeta() {
      return this.blockchain.blockMetaInfo
    },
    pageBlockTitle() {
      if (this.block && this.block.header) {
        return "Block #" + num.prettyInt(this.block.header.height)
      } else {
        return "Loading..."
      }
    },
    nextBlockAvailable() {
      return (
        this.block &&
        this.block.header &&
        this.block.header.height < this.blockchainHeight
      )
    }
  },
  methods: {
    fetchBlock() {
      this.$store.dispatch("getBlock", parseInt(this.$route.params.block))
    }
  },
  mounted() {
    this.fetchBlock()
  },
  watch: {
    $route() {
      this.fetchBlock()
    }
  }
}
</script>
