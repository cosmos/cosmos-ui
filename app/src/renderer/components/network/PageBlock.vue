<template>
  <tm-page data-title="Block">
    <tm-data-error v-if="!connected || !block" />
    <template v-else>
      <div class="page-profile__header page-profile__section block">
        <div class="row">
          <div class="page-profile__header__info">
            <div class="page-profile__status-and-title">
              <h2 class="page-profile__title">
                Block {{ blockTitle }}
              </h2>
              <h3 class="page-profile__subtitle">
                {{ block.block_meta.block_id ? block.block_meta.block_id.hash : '' }}
              </h3>
            </div>
          </div>
        </div>

        <div class="row">
          <dl class="info_dl colored_dl">
            <dt>Time</dt>
            <dd>{{ blockTime }}</dd>
          </dl>
        </div>
      </div>

      <div class="page-profile__section block">
        <div class="row">
          <div class="column">
            <dl class="info_dl colored_dl">
              <dt>Transactions</dt>
              <dd>{{ block.block && block.block.data.txs || `No Transactions` }}</dd>
            </dl>
          </div>
        </div>
      </div>

      <div class="page-profile__section block">
        <div class="row">
          <div class="column">
            <dl class="info_dl colored_dl">
              <dt>Evidence</dt>
              <dd>{{ block.block.evidence.evidence || `No Evidence` }}</dd>
            </dl>
          </div>
        </div>
      </div>
    </template>
  </tm-page>
</template>

<script>
import moment from "moment"
import { mapGetters } from "vuex"
import num from "scripts/num"
import TmDataError from "common/TmDataError"
import TmPage from "common/TmPage"
export default {
  name: `page-block`,
  components: {
    TmDataError,
    TmPage
  },
  data: () => ({
    num,
    moment
  }),
  computed: {
    ...mapGetters([`connected`, `block`, `lastHeader`]),
    properties() {
      return [
        {
          title: `Proposer`
        },
        {
          title: `Time`
        },
        {
          title: `Round`
        }
      ]
    },
    blockTitle({ num, block } = this) {
      if (!block.block) return `n/a`
      return `#` + num.prettyInt(block.block.header.height)
    },
    blockTime({ moment, block } = this) {
      if (!block.block) return `n/a`
      return moment(block.block.header.time).format(`MMM Do YYYY, HH:mm:ss`)
    }
  },
  watch: {
    $route: function() {
      this.getBlock()
    }
  },
  mounted() {
    this.getBlock()
  },
  methods: {
    async getBlock({ $store, $route, $router, lastHeader } = this) {
      // query first for the block so we don't fail if the user started from this route and hasn't received any lastHeader yet
      const blockInfo = await $store.dispatch(
        `queryBlockInfo`,
        $route.params.height
      )

      if (
        !blockInfo &&
        Number($route.params.height) > Number(lastHeader.height)
      ) {
        $router.push(`/404`)
        return
      }
    }
  }
}
</script>
