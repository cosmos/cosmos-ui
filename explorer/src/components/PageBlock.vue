<template lang="pug">
page(:title="`Block ${block.header.height}`")
  tool-bar
    router-link(to="/"): i.material-icons arrow_back
    a(:href="blockUrl" target="_blank") JSON
    router-link(:to="{ name: 'block', params: { block: block.header.height - 1 }}"): i.material-icons chevron_left
    router-link(:to="{ name: 'block', params: { block: block.header.height + 1 }}"): i.material-icons chevron_right

  part(title='Header')
    list-item(dt="Chain ID" :dd="block.header.chain_id")
    list-item(dt="Time" :dd="block.header.time")
    list-item(dt="Transactions" :dd="block.header.num_txs")
    list-item(dt="Last Commit Hash" :dd="block.header.last_commit_hash")
    list-item(dt="Validators Hash" :dd="block.header.validators_hash")
    list-item(dt="App Hash" :dd="block.header.app_hash")

  part(title='Last Block')
    list-item(dt="Hash" :dd="block.header.last_block_id.hash")
    list-item(dt="Parts Total"
      :dd="block.header.last_block_id.parts.total")
    list-item(dt="Parts Hash" :dd="block.header.last_block_id.parts.hash")

  part(title="Precommit"
    v-for="p in block.last_commit.precommits"
    :key="p.validator_address" v-if="p !== null")
    list-item(dt="Address" :dd="p.validator_address")
    list-item(dt="Index" :dd="p.validator_index")
    list-item(dt="Round" :dd="p.round")
    list-item(:dt="`Sig (${p.signature.type})`"
    :dd="p.signature.data")

  part(title='Transactions')
  part(v-for="tx in block.data.txs" :title="tx.hash" :key="tx.hash")
    list-item(v-for="(tx, key) in tx" :key="tx.hash + 'key'" :dt="key" :dd="tx")
</template>

<script>
import { mapGetters } from 'vuex'
import axios from 'axios'
import createHash from 'create-hash'
import varint from 'varint'
import b64 from 'base64-js'
import ToolBar from './NiToolBar'
import ListItem from './NiListItem'
import Part from './NiPart'
import Page from './NiPage'
export default {
  name: 'page-block',
  components: {
    ToolBar,
    ListItem,
    Part,
    Page
  },
  computed: {
    ...mapGetters(['bc'])
  },
  data: () => ({
    blockUrl: '',
    block_meta: {
      block_id: {
        hash: '',
        parts: {
          total: 0,
          hash: ''
        }
      },
      header: {
        chain_id: '',
        height: 0,
        time: '',
        num_txs: 0,
        last_block_id: {
          hash: '',
          parts: {
            total: 0,
            hash: ''
          }
        },
        last_commit_hash: '',
        data_hash: '',
        validators_hash: '',
        app_hash: ''
      }
    },
    block: {
      header: {
        chain_id: '',
        height: 0,
        time: '',
        num_txs: 0,
        last_block_id: {
          hash: '',
          parts: {
            total: 0,
            hash: ''
          }
        },
        last_commit_hash: '',
        data_hash: '',
        validators_hash: '',
        app_hash: ''
      },
      data: {
        txs: []
      },
      last_commit: {
        blockID: {
          hash: '',
          parts: {
            total: 0,
            hash: ''
          }
        },
        precommits: [
          {
            validator_address: '',
            validator_index: 0,
            height: 0,
            round: 0,
            type: 0,
            block_id: {
              hash: '',
              parts: {
                total: 0,
                hash: ''
              }
            },
            signature: [0, '']
          }
        ]
      }
    }
  }),
  methods: {
    async fetchBlock () {
      let url = `${this.bc.url}/block?height=${this.$route.params.block}`
      let json = await axios.get(url)
      this.block_meta = json.data.result.block_meta
      this.block = json.data.result.block
      this.block.data.txs && this.queryTxs().catch((err) => console.log(err))
    },
    queryTxs () {
      return this.queryTx(this.block.data.txs.length)
    },
    queryTx (len, key = 0) {
      return new Promise((resolve, reject) => {
        if (key >= len) resolve()
        let txstring = atob(this.block.data.txs[key])
        // console.log(txstring)
        let txbytes = b64.toByteArray(this.block.data.txs[key])
        // console.log(txbytes)
        let varintlen = new Uint8Array(varint.encode(txbytes.length))
        // console.log(varintlen)
        let tmp = new Uint8Array(varintlen.byteLength + txbytes.byteLength)
        tmp.set(new Uint8Array(varintlen), 0)
        tmp.set(new Uint8Array(txbytes), varintlen.byteLength)
        // console.log(tmp)
        let hash = createHash('ripemd160').update(Buffer.from(tmp)).digest('hex')
        // console.log(hash)
        let url = `${this.bc.url}/tx?hash=0x${hash}`
        axios.get(url).then(json => {
          // console.log(json)
          json.data.result.string = txstring
          this.block.data.txs.splice(key, 1, json.data.result)
          this.queryTx(len, key + 1).then(resolve).catch(reject)
        }).catch(reject)
      })
    }
  },
  async mounted () {
    await this.fetchBlock()
  },
  watch: {
    '$route' (to, from) {
      this.fetchBlock()
    }
  }
}
</script>
