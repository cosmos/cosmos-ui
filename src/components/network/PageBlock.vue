<template>
  <TmPage data-title="Block">
    <TmDataError v-if="!connected || !block" />
    <template v-else>
      <div class="page-profile__header page-profile__section block">
        <div class="row">
          <div class="page-profile__header__info">
            <div class="page-profile__status-and-title">
              <h2 class="page-profile__title">
                Block {{ blockTitle || `--` }}
              </h2>
            </div>
          </div>
        </div>

        <div class="row">
          <dl class="info_dl colored_dl">
            <dt>Time</dt>
            <dd>{{ blockTime }}</dd>
          </dl>
        </div>
      </div>

      <div class="page-profile__section block">
        <div class="row">
          <div class="column">
            <h3 v-if="block.transactions" class="page-profile__section-title">
              Transactions
            </h3>
            <TmDataMsg
              v-if="block.transactions && block.transactions.length === 0"
              icon="info_outline"
            >
              <div slot="title">
                No Transactions
              </div>
              <div slot="subtitle">
                This block doesn't contain any transactions.
              </div>
            </TmDataMsg>
            <LiAnyTransaction
              v-for="tx in block.transactions"
              :key="tx.txhash"
              :validators="delegates.delegates"
              validators-url="/staking/validators"
              proposals-url="/governance"
              :transaction="tx"
              :address="session.address || ``"
              :bonding-denom="bondDenom"
              :height="block.block.header.height"
              :time="block.block.header.time"
              :unbonding-time="
                getUnbondingTime(tx, delegation.unbondingDelegations)
              "
            />
            <br />
          </div>
        </div>
      </div>
    </template>
  </TmPage>
</template>

<script>
import moment from "moment"
import { mapGetters } from "vuex"
import num from "scripts/num"
import { getUnbondingTime } from "scripts/time"
import TmDataError from "common/TmDataError"
import TmPage from "common/TmPage"
import LiAnyTransaction from "transactions/LiAnyTransaction"
import TmDataMsg from "common/TmDataMsg"
export default {
  name: `page-block`,
  components: {
    TmDataError,
    TmDataMsg,
    TmPage,
    LiAnyTransaction
  },
  data: () => ({
    num,
    moment,
    getUnbondingTime
  }),
  computed: {
    ...mapGetters([
      `connected`,
      `block`,
      `bondDenom`,
      `lastHeader`,
      `delegates`,
      `delegation`,
      `session`
    ]),
    properties() {
      return [
        {
          title: `Proposer`
        },
        {
          title: `Time`
        },
        {
          title: `Round`
        }
      ]
    },
    blockTitle({ num, block } = this) {
      if (!block.block) return `--`
      return `#` + num.prettyInt(block.block.header.height)
    },
    blockTime({ moment, block } = this) {
      if (!block.block) return `--`
      return moment(block.block.header.time).format(`MMM Do YYYY, HH:mm:ss`)
    }
  },
  watch: {
    "$route.params.height": async function() {
      await this.getBlock()
    }
  },
  async mounted() {
    await this.getBlock()
  },
  methods: {
    async getBlock({ $store, $route, $router, lastHeader } = this) {
      // query first for the block so we don't fail if the user started from this route and hasn't received any lastHeader yet
      await $store.dispatch(`queryBlockInfo`, $route.params.height)

      if (!this.block && $route.params.height > lastHeader.height) {
        $router.push(`/404`)
        return
      }
      await $store.dispatch(`getBlockTxs`, $route.params.height)
    }
  }
}
</script>
