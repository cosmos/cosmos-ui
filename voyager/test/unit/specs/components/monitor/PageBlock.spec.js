import setup from "../../../helpers/vuex-setup"
import htmlBeautify from "html-beautify"
import PageBlock from "renderer/components/monitor/PageBlock"

describe("PageBlock", () => {
  let wrapper, store
  let { mount } = setup()

  beforeEach(() => {
    let instance = mount(PageBlock, {
      getters: {
        blockchain: () => ({
          blocks: [
            {
              header: {
                height: 100000
              }
            }
          ],
          block: {
            header: {
              last_block_id: {
                hash: "last-hash",
                parts: { total: 0 }
              },
              num_txs: 0,
              height: 10,
              time: 1608
            },
            last_commit: { precommits: [] },
            data: { txs: 0 }
          },
          blockMetaInfo: { block_id: { hash: "hash" } }
        })
      }
    })
    wrapper = instance.wrapper
    store = instance.store

    wrapper.update()
  })

  it("has the expected html structure", () => {
    expect(htmlBeautify(wrapper.html())).toMatchSnapshot()
  })

  it("should dispatch getBlock when mounted", () => {
    expect(store.dispatch).toHaveBeenCalledWith("getBlock", NaN) // NaN as it trys to parse $route.param.block which is not set
  })

  it("should show a loading state if loading", () => {
    let { wrapper } = mount(PageBlock, {
      getters: {
        blockchain: () => ({
          blocks: [
            {
              header: {
                height: 100000
              }
            }
          ],
          block: {},
          blockMetaInfo: { block_id: {} },
          blockLoading: true
        })
      },
      stubs: { "data-loading": "<data-loading />" }
    })
    wrapper.update()

    expect(wrapper.vm.block).toEqual({})
    expect(wrapper.contains("data-loading")).toBe(true)
    expect(htmlBeautify(wrapper.html())).toMatchSnapshot()
  })

  it("should survive no blocks being available", () => {
    let { wrapper } = mount(PageBlock, {
      getters: {
        blockchain: () => ({
          blocks: [],
          block: {},
          blockMetaInfo: {},
          blockLoading: true
        })
      },
      stubs: { "data-loading": "<data-loading />" }
    })
    wrapper.update()

    expect(wrapper.contains("data-loading")).toBe(true)
  })

  it("should disable the next block button if last block", () => {
    let { wrapper } = mount(PageBlock, {
      getters: {
        blockchain: () => ({
          blocks: [
            {
              header: {
                height: 10
              }
            }
          ],
          block: {
            header: {
              last_block_id: {
                hash: "last-hash",
                parts: { total: 0 }
              },
              num_txs: 0,
              height: 10,
              time: 1608
            },
            last_commit: { precommits: [] },
            data: { txs: 0 }
          },
          blockMetaInfo: { block_id: {} },
          blockLoading: true
        })
      }
    })
    expect(wrapper.vm.nextBlockAvailable).toBe(false)
    expect(wrapper.$el).toMatchSnapshot()
  })
})
