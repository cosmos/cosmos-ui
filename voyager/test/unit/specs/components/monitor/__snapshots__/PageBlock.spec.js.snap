// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageBlock has the expected html structure 1`] = `
"<div class=\\"ni-page\\">
   <header class=\\"ni-page-header\\">
      <div class=\\"ni-page-header-container\\">
         <div class=\\"ni-page-header-text\\">
            <div class=\\"ni-page-header-title\\">
               <h2>
                  Block #10
               </h2>
</div>
   <div class=\\"ni-page-header-subtitle\\">
      <h3></h3>
</div>
</div>
   <menu class=\\"ni-page-header-menu\\">
      <div>
         <div>
            <div class=\\"ni-tool-bar\\">
               <div class=\\"ni-tool-bar-container\\">
                  <div class=\\"main\\">
                     <a href=\\"#/blocks/9\\" class=\\"\\"><i class=\\"material-icons\\">chevron_left</i></a><a href=\\"#/blocks/11\\" class=\\"\\"><i class=\\"material-icons\\">chevron_right</i></a>
                  </div>
      <a class=\\"back\\"><i class=\\"material-icons\\">arrow_back</i></a><a class=\\"help\\"><i class=\\"material-icons\\">help_outline</i></a><a class=\\"sign-out\\"><i class=\\"material-icons\\">exit_to_app</i></a>
   </div>
</div>
</div>
</div>
</menu>
</div>
</header>
<main class=\\"ni-page-main ps\\">
   <section class=\\"ni-part\\">
      <div class=\\"ni-part-container\\">
         <!---->
         <main class=\\"ni-part-main\\">
            <div class=\\"ni-li\\">
               <div class=\\"ni-li-container\\">
                  <div class=\\"ni-li-thumb\\">
                     <!---->
                  </div>
         <div class=\\"ni-li-dl\\">
            <div class=\\"ni-li-dt\\">
               Block Hash
            </div>
         <div class=\\"ni-li-dd\\">
            hash
         </div>
</div>
</div>
</div>
</main>
</div>
</section>
<section class=\\"ni-part\\">
   <div class=\\"ni-part-container\\">
      <header class=\\"ni-part-header\\">
         <div class=\\"ni-part-header-container\\">
            <div class=\\"ni-part-title h5\\">
               Header
            </div>
      <menu class=\\"ni-part-menu\\"></menu>
</div>
</header>
   <main class=\\"ni-part-main\\">
      <div class=\\"ni-li\\">
         <div class=\\"ni-li-container\\">
            <div class=\\"ni-li-thumb\\">
               <!---->
            </div>
      <div class=\\"ni-li-dl\\">
         <div class=\\"ni-li-dt\\">
            Chain ID
         </div>
      <div class=\\"ni-li-dd\\"></div>
</div>
</div>
</div>
      <div class=\\"ni-li\\">
         <div class=\\"ni-li-container\\">
            <div class=\\"ni-li-thumb\\">
               <!---->
            </div>
      <div class=\\"ni-li-dl\\">
         <div class=\\"ni-li-dt\\">
            Time
         </div>
      <div class=\\"ni-li-dd\\">
         January 1st 1970 â€” 02:00:01
      </div>
</div>
</div>
</div>
      <div class=\\"ni-li\\">
         <div class=\\"ni-li-container\\">
            <div class=\\"ni-li-thumb\\">
               <!---->
            </div>
      <div class=\\"ni-li-dl\\">
         <div class=\\"ni-li-dt\\">
            Transactions
         </div>
      <div class=\\"ni-li-dd\\">
         0
      </div>
</div>
</div>
</div>
      <div class=\\"ni-li\\">
         <div class=\\"ni-li-container\\">
            <div class=\\"ni-li-thumb\\">
               <!---->
            </div>
      <div class=\\"ni-li-dl\\">
         <div class=\\"ni-li-dt\\">
            Last Commit Hash
         </div>
      <div class=\\"ni-li-dd\\"></div>
</div>
</div>
</div>
      <div class=\\"ni-li\\">
         <div class=\\"ni-li-container\\">
            <div class=\\"ni-li-thumb\\">
               <!---->
            </div>
      <div class=\\"ni-li-dl\\">
         <div class=\\"ni-li-dt\\">
            Validators Hash
         </div>
      <div class=\\"ni-li-dd\\"></div>
</div>
</div>
</div>
      <div class=\\"ni-li\\">
         <div class=\\"ni-li-container\\">
            <div class=\\"ni-li-thumb\\">
               <!---->
            </div>
      <div class=\\"ni-li-dl\\">
         <div class=\\"ni-li-dt\\">
            App Hash
         </div>
      <div class=\\"ni-li-dd\\"></div>
</div>
</div>
</div>
</main>
</div>
</section>
<section class=\\"ni-part\\">
   <div class=\\"ni-part-container\\">
      <header class=\\"ni-part-header\\">
         <div class=\\"ni-part-header-container\\">
            <div class=\\"ni-part-title h5\\">
               Last Block
            </div>
      <menu class=\\"ni-part-menu\\"></menu>
</div>
</header>
   <main class=\\"ni-part-main\\">
      <div class=\\"ni-li\\">
         <div class=\\"ni-li-container\\">
            <div class=\\"ni-li-thumb\\">
               <!---->
            </div>
      <div class=\\"ni-li-dl\\">
         <div class=\\"ni-li-dt\\">
            Hash
         </div>
      <div class=\\"ni-li-dd\\">
         last-hash
      </div>
</div>
</div>
</div>
      <div class=\\"ni-li\\">
         <div class=\\"ni-li-container\\">
            <div class=\\"ni-li-thumb\\">
               <!---->
            </div>
      <div class=\\"ni-li-dl\\">
         <div class=\\"ni-li-dt\\">
            Parts Total
         </div>
      <div class=\\"ni-li-dd\\">
         0
      </div>
</div>
</div>
</div>
      <div class=\\"ni-li\\">
         <div class=\\"ni-li-container\\">
            <div class=\\"ni-li-thumb\\">
               <!---->
            </div>
      <div class=\\"ni-li-dl\\">
         <div class=\\"ni-li-dt\\">
            Parts Hash
         </div>
      <div class=\\"ni-li-dd\\"></div>
</div>
</div>
</div>
</main>
</div>
</section>
<section class=\\"ni-part\\">
   <div class=\\"ni-part-container\\">
      <header class=\\"ni-part-header\\">
         <div class=\\"ni-part-header-container\\">
            <div class=\\"ni-part-title h5\\">
               Transactions
            </div>
      <menu class=\\"ni-part-menu\\"></menu>
</div>
</header>
   <main class=\\"ni-part-main\\">
      <div class=\\"ni-data-msg\\">
         <div class=\\"ni-data-msg__icon\\">
            <i class=\\"material-icons\\">info_outline</i>
         </div>
      <div class=\\"ni-data-msg__text\\">
         <div class=\\"ni-data-msg__title\\">
            <h4>
               Empty Block
            </h4>
</div>
      <div class=\\"ni-data-msg__subtitle\\">
         <h5>
            There were no transactions in this block.
         </h5>
</div>
</div>
</div>
</main>
</div>
</section>
<div class=\\"ps__rail-x\\" style=\\"left: 0px; top: 0px;\\">
   <div class=\\"ps__thumb-x\\" tabindex=\\"0\\" style=\\"left: 0px; width: 0px;\\"></div>
</div>
<div class=\\"ps__rail-y\\" style=\\"top: 0px; left: 0px;\\">
   <div class=\\"ps__thumb-y\\" tabindex=\\"0\\" style=\\"top: 0px; height: 0px;\\"></div>
</div>
</main>
</div>"
`;

exports[`PageBlock should disable the next block button if last block 1`] = `undefined`;

exports[`PageBlock should show a loading state if loading 1`] = `
"<div class=\\"ni-page\\">
   <header class=\\"ni-page-header\\">
      <div class=\\"ni-page-header-container\\">
         <div class=\\"ni-page-header-text\\">
            <div class=\\"ni-page-header-title\\">
               <h2>
                  Loading...
               </h2>
</div>
   <div class=\\"ni-page-header-subtitle\\">
      <h3></h3>
</div>
</div>
   <menu class=\\"ni-page-header-menu\\">
      <div></div>
</menu>
</div>
</header>
<main class=\\"ni-page-main\\">
   <data-loading></data-loading>
</main>
</div>"
`;
